source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.cc

```diff
@@ -32,13 +32,23 @@ Http::FilterHeadersStatus AdaptiveConcurrencyFilter::decodeHeaders(Http::HeaderM
     return Http::FilterHeadersStatus::StopIteration;
   }
 
-  rq_start_time_ = config_->timeSource().monotonicTime();
+  // When the deferred_sample_task_ object is destroyed, the time difference between its destruction
+  // and the request start time is measured as the request latency. This value is sampled by the
+  // concurrency controller either when encoding is complete or during destruction of this filter
+  // object.
+  const auto rq_start_time = config_->timeSource().monotonicTime();
+  deferred_sample_task_ = std::make_unique<Cleanup>([this, rq_start_time]() {
```

Comment:
In the reset case, I don't think you want to actually record a latency sample, you just want to decrement the active request count, so I think you need to account for that.

Also, I would do explicit cleanup on `onDestroy()` to avoid any issues around deferred deletion. Basically, in the `onDestroy()` case if there is still an active count, you should just decrement the count and not store a sample.

IMO it's cleaner the controller to return an interface object that hides all of this. It would have a single method to record a latency sample, and on destruction would decrement the count. But this can work too, with all callers having to deal with this accounting.